axon-cosmosdb/src/main/java/org/axonframework/extensions/cosmosdb/eventsourcing/tokenstore/CosmosTokenStore.java

```diff
@@ -0,0 +1,516 @@
+        }
+    }
+
+    @Override
+    public void releaseClaim(@Nonnull String processorName, int segment) {
+        getCurrent(processorName, segment)
+                .flatMap(current -> {
+                    validateCurrent(current.getItem());
+                    return getContainer(processorName).replaceItem(
+                            current.getItem().release(),
+                            current.getItem().getId(),
+                            new PartitionKey(current.getItem().getId()),
+                            new CosmosItemRequestOptions().setIfMatchETag(current.getETag())
+                    );
+                })
+                .block();
+    }
+
+    @Override
+    public void initializeSegment(@Nullable TrackingToken token, @Nonnull String processorName, int segment)
+            throws UnableToInitializeTokenException {
+        try {
+            getContainer(processorName).createItem(
+                    createInitialItem(token, segment),
+                    new PartitionKey(Integer.toString(segment)),
+                    new CosmosItemRequestOptions().setIfMatchETag(null)
+            ).block();
+        } catch (Exception e) {
+            throw new UnableToInitializeTokenException("Could not initialize token.", e);
+        }
+    }
+
+    @Override
+    public void deleteToken(@Nonnull String processorName, int segment) throws UnableToClaimTokenException {
+        try {
+            getCurrent(processorName, segment)
+                    .flatMap(current -> {
+                        String owner = current.getItem().getOwner();
+                        if (!nodeId.equals(owner)) {
+                            return Mono.error(new UnableToClaimTokenException(
+                                    String.format(CANT_DELETE_MESSAGE_FORMAT, owner, nodeId)));
+                        }
+                        return getContainer(processorName).deleteAllItemsByPartitionKey(
+                                new PartitionKey(Integer.toString(segment)),
+                                new CosmosItemRequestOptions().setIfMatchETag(null)
+                        );
+                    })
+                    .block();
+        } catch (UnableToClaimTokenException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new UnableToClaimTokenException("Could not delete token.", e);
+        }
+    }
+
+    @Override
+    public boolean requiresExplicitSegmentInitialization() {
+        return true;
+    }
+
+    @Override
+    public int[] fetchSegments(@Nonnull String processorName) {
+        List<Integer> segments = getContainer(processorName)
+                .queryItems(ALL_QUERY, CosmosTokenItem.class)
+                .map(CosmosTokenItem::getSegment)
+                .collectList()
+                .block();
+        if (isNull(segments)) {
+            return new int[0];
+        }
+        // toArray doesn't work because of autoboxing limitations
```

Comment:
I see this is a copy from the `MongoTokenStore`. The `JpaTokenStore` does the following on the `List` it receives, with I assume success:

```java
final List<Integer> resultList = ...;
resultList.stream().mapToInt(i -> i).toArray();
```

Granted, both aren't very pretty.

Reply:
Changed.